File: haskell/strain/src/Strain.hs
module Strain (keep, discard) where

discard :: (a -> Bool) -> [a] -> [a]
discard f = keep (not . f)

keep :: (a -> Bool) -> [a] -> [a]
keep f = foldr g []
  where
    g x xs = if f x then x : xs else xs

File: scala/strain/src/main/scala/Strain.scala
object Strain {
  def keep[A](xs : Seq[A], f: A => Boolean) =
    (xs :\ Seq.empty[A]){ (x, ys) =>
      if( f(x) ) x +: ys else ys
    }

  val not: Boolean => Boolean = !_

  def discard[A](xs : Seq[A], f: A => Boolean) =
    keep(xs, not compose f)
}
